File: web/src/ChatTemplate.ts
import { ChatMessage, ChatRole } from './ChatTypes';

/**
 * Chat template formats for different model families
 */
export enum ChatTemplate {
  /** Llama 3 chat format */
  LLAMA3 = 'llama3',

  /** ChatML format (used by OpenAI, GPT-4, etc.) */
  CHATML = 'chatml',

  /** Mistral/Mixtral chat format */
  MISTRAL = 'mistral',

  /** Generic markdown-style format (### System: / ### User: / ### Assistant:) */
  GENERIC = 'generic',

  /** Qwen3 format — ChatML base with tool-call support */
  QWEN3 = 'qwen3',

  /** Gemma3 format — <start_of_turn> / <end_of_turn> with "model" role */
  GEMMA3 = 'gemma3',
}

/**
 * Formats messages according to a chat template
 */
export function formatMessages(template: ChatTemplate, messages: ChatMessage[]): string {
  switch (template) {
    case ChatTemplate.LLAMA3:
      return formatLlama3(messages);
    case ChatTemplate.CHATML:
      return formatChatML(messages);
    case ChatTemplate.MISTRAL:
      return formatMistral(messages);
    case ChatTemplate.GENERIC:
      return formatGeneric(messages);
    case ChatTemplate.QWEN3:
      return formatQwen3(messages);
    case ChatTemplate.GEMMA3:
      return formatGemma3(messages);
  }
}

function formatLlama3(messages: ChatMessage[]): string {
  let prompt = '<|begin_of_text|>';

  for (const message of messages) {
    prompt += `<|start_header_id|>${message.role}<|end_header_id|>\n\n`;
    prompt += `${message.content}<|eot_id|>`;
  }

  // Add assistant prompt to continue generation
  prompt += '<|start_header_id|>assistant<|end_header_id|>\n\n';

  return prompt;
}

function formatChatML(messages: ChatMessage[]): string {
  let prompt = '';

  for (const message of messages) {
    prompt += `<|im_start|>${message.role}\n`;
    prompt += `${message.content}<|im_end|>\n`;
  }

  // Add assistant prompt to continue generation
  prompt += '<|im_start|>assistant\n';

  return prompt;
}

function formatMistral(messages: ChatMessage[]): string {
  let prompt = '<s>';
  let pendingSystem: string | undefined;

  for (const message of messages) {
    switch (message.role) {
      case ChatRole.SYSTEM:
        pendingSystem = message.content;
        break;
      case ChatRole.USER:
        prompt += '[INST]';
        if (pendingSystem !== undefined) {
          prompt += ` <<SYS>>\n${pendingSystem}\n<</SYS>>\n\n`;
          pendingSystem = undefined;
        } else {
          prompt += ' ';
        }
        prompt += `${message.content} [/INST]`;
        break;
      case ChatRole.ASSISTANT:
        prompt += ` ${message.content}</s>`;
        break;
      case ChatRole.TOOL_RESULT:
        prompt += `[INST] Tool result: ${message.content} [/INST]`;
        break;
      case ChatRole.SUMMARY:
        // Inject summary as the next system block
        pendingSystem = `[Context Summary] ${message.content}`;
        break;
      // TOOL_CALL is already part of the assistant output — skip to avoid duplication
    }
  }

  return prompt;
}

function formatGeneric(messages: ChatMessage[]): string {
  let prompt = '';

  for (const message of messages) {
    switch (message.role) {
      case ChatRole.SYSTEM:
        prompt += `### System:\n${message.content}\n\n`;
        break;
      case ChatRole.USER:
        prompt += `### User:\n${message.content}\n`;
        break;
      case ChatRole.ASSISTANT:
        prompt += `### Assistant:\n${message.content}\n`;
        break;
      case ChatRole.TOOL_CALL:
        prompt += `### Tool Call:\n${message.content}\n`;
        break;
      case ChatRole.TOOL_RESULT:
        prompt += `### Tool Result:\n${message.content}\n`;
        break;
      case ChatRole.SUMMARY:
        prompt += `### Context Summary:\n${message.content}\n\n`;
        break;
    }
  }

  prompt += '### Assistant:\n';
  return prompt;
}

function formatQwen3(messages: ChatMessage[]): string {
  let prompt = '';

  for (const message of messages) {
    switch (message.role) {
      case ChatRole.SYSTEM:
        prompt += `<|im_start|>system\n${message.content}<|im_end|>\n`;
        break;
      case ChatRole.USER:
        prompt += `<|im_start|>user\n${message.content}<|im_end|>\n`;
        break;
      case ChatRole.ASSISTANT:
        prompt += `<|im_start|>assistant\n${message.content}<|im_end|>\n`;
        break;
      case ChatRole.TOOL_CALL:
        // Hermes-style: tool call is part of the assistant turn
        prompt += `<|im_start|>assistant\n${message.content}<|im_end|>\n`;
        break;
      case ChatRole.TOOL_RESULT:
        // Hermes-style: tool results use the "tool" role
        prompt += `<|im_start|>tool\n${message.content}<|im_end|>\n`;
        break;
      case ChatRole.SUMMARY:
        prompt += `<|im_start|>system\n[Context Summary] ${message.content}<|im_end|>\n`;
        break;
    }
  }

  // Prime assistant turn for generation
  prompt += '<|im_start|>assistant\n';
  return prompt;
}

function formatGemma3(messages: ChatMessage[]): string {
  let prompt = '';

  for (const message of messages) {
    switch (message.role) {
      case ChatRole.SYSTEM:
        // Gemma3 has no dedicated system turn — prepend to first user turn
        prompt += `<start_of_turn>user\n${message.content}<end_of_turn>\n`;
        break;
      case ChatRole.USER:
        prompt += `<start_of_turn>user\n${message.content}<end_of_turn>\n`;
        break;
      case ChatRole.ASSISTANT:
        // Gemma3 uses "model" as the assistant role name
        prompt += `<start_of_turn>model\n${message.content}<end_of_turn>\n`;
        break;
      case ChatRole.TOOL_CALL:
        prompt += `<start_of_turn>model\n${message.content}<end_of_turn>\n`;
        break;
      case ChatRole.TOOL_RESULT:
        prompt += `<start_of_turn>tool\n${message.content}<end_of_turn>\n`;
        break;
      case ChatRole.SUMMARY:
        prompt += `<start_of_turn>user\n[Context Summary] ${message.content}<end_of_turn>\n`;
        break;
    }
  }

  // Prime model turn for generation
  prompt += '<start_of_turn>model\n';
  return prompt;
}

File: kotlin/src/main/kotlin/com/edgeveda/sdk/ChatTemplate.kt
package com.edgeveda.sdk

/**
 * Chat template formats for different model families
 */
enum class ChatTemplate {
    /** Llama 3 chat format */
    LLAMA3,

    /** ChatML format (used by OpenAI, GPT-4, etc.) */
    CHATML,

    /** Mistral/Mixtral chat format */
    MISTRAL,

    /** Generic ### System/User/Assistant format */
    GENERIC,

    /** Qwen3 ChatML format (same tokens as CHATML, kept separate for future tool-calling) */
    QWEN3,

    /** Gemma 3 format — system merged into first user turn; "model" role for assistant */
    GEMMA3;

    /**
     * Formats a list of messages according to the template
     * @param messages The messages to format
     * @return The formatted prompt string
     */
    fun format(messages: List<ChatMessage>): String {
        return when (this) {
            LLAMA3 -> formatLlama3(messages)
            CHATML -> formatChatML(messages)
            MISTRAL -> formatMistral(messages)
            GENERIC -> formatGeneric(messages)
            QWEN3 -> formatQwen3(messages)
            GEMMA3 -> formatGemma3(messages)
        }
    }
    
    private fun formatLlama3(messages: List<ChatMessage>): String {
        val prompt = StringBuilder()

        // Required BOS token for Llama 3 chat format
        prompt.append("<|begin_of_text|>")

        for (message in messages) {
            prompt.append("<|start_header_id|>${message.role.value}<|end_header_id|>\n\n")
            prompt.append("${message.content}<|eot_id|>")
        }

        // Add assistant prompt to continue generation
        prompt.append("<|start_header_id|>assistant<|end_header_id|>\n\n")

        return prompt.toString()
    }
    
    private fun formatChatML(messages: List<ChatMessage>): String {
        val prompt = StringBuilder()
        
        for (message in messages) {
            prompt.append("<|im_start|>${message.role.value}\n")
            prompt.append("${message.content}<|im_end|>\n")
        }
        
        // Add assistant prompt to continue generation
        prompt.append("<|im_start|>assistant\n")
        
        return prompt.toString()
    }
    
    private fun formatMistral(messages: List<ChatMessage>): String {
        val prompt = StringBuilder()

        // Mistral requires the system prompt to be embedded inside the first
        // [INST] block using <<SYS>> markers, not placed before it as bare text.
        val systemText = messages.firstOrNull { it.role == ChatRole.SYSTEM }?.content
        val conversationMessages = messages.filter { it.role != ChatRole.SYSTEM }

        var isFirstUserMessage = true
        for (message in conversationMessages) {
            when (message.role) {
                ChatRole.USER -> {
                    val content = if (isFirstUserMessage && systemText != null) {
                        "<<SYS>>\n$systemText\n<</SYS>>\n\n${message.content}"
                    } else {
                        message.content
                    }
                    prompt.append("<s>[INST] $content [/INST]")
                    isFirstUserMessage = false
                }
                ChatRole.ASSISTANT -> {
                    prompt.append(" ${message.content}</s>")
                }
                ChatRole.SYSTEM -> { /* handled above */ }
            }
        }

        return prompt.toString()
    }

    private fun formatGeneric(messages: List<ChatMessage>): String {
        val prompt = StringBuilder()
        for (msg in messages) {
            when (msg.role) {
                ChatRole.SYSTEM    -> prompt.append("### System:\n${msg.content}\n\n")
                ChatRole.USER      -> prompt.append("### User:\n${msg.content}\n")
                ChatRole.ASSISTANT -> prompt.append("### Assistant:\n${msg.content}\n")
            }
        }
        prompt.append("### Assistant:\n")
        return prompt.toString()
    }

    private fun formatQwen3(messages: List<ChatMessage>): String {
        val prompt = StringBuilder()
        for (msg in messages) {
            when (msg.role) {
                ChatRole.SYSTEM    -> prompt.append("<|im_start|>system\n${msg.content}<|im_end|>\n")
                ChatRole.USER      -> prompt.append("<|im_start|>user\n${msg.content}<|im_end|>\n")
                ChatRole.ASSISTANT -> prompt.append("<|im_start|>assistant\n${msg.content}<|im_end|>\n")
            }
        }
        prompt.append("<|im_start|>assistant\n")
        return prompt.toString()
    }

    private fun formatGemma3(messages: List<ChatMessage>): String {
        val prompt = StringBuilder()
        var pendingSystem: String? = null

        for (msg in messages) {
            when (msg.role) {
                ChatRole.SYSTEM -> pendingSystem = msg.content
                ChatRole.USER -> {
                    val content = if (pendingSystem != null) {
                        "$pendingSystem\n\n${msg.content}".also { pendingSystem = null }
                    } else {
                        msg.content
                    }
                    prompt.append("<start_of_turn>user\n${content}<end_of_turn>\n")
                }
                ChatRole.ASSISTANT ->
                    prompt.append("<start_of_turn>model\n${msg.content}<end_of_turn>\n")
            }
        }

        prompt.append("<start_of_turn>model\n")
        return prompt.toString()
    }
}